snuba/migrations/migration.py

```diff
@@ -44,6 +44,55 @@ def forwards(self, context: Context, dry_run: bool) -> None:
     def backwards(self, context: Context, dry_run: bool) -> None:
         raise NotImplementedError
 
+    @abstractmethod
+    def forwards_local(self) -> Sequence[Operation]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def backwards_local(self) -> Sequence[Operation]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def forwards_dist(self) -> Sequence[Operation]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def backwards_dist(self) -> Sequence[Operation]:
+        raise NotImplementedError
```

Comment:
Having both the forwards method and forwards_local/dist in the same class and all abstract does not seem to fit the design of the multi step migrations since the local/dist methods are supposed to be invoked directly by the forwards/backwards methods and none is implemented to enforce this design constraint.

It seems the problem you are trying to address is the one of making the initial migration a multi step one to have the support of dry run.
I think the main issue here is that the initial migration is looking more and more a multi step migration and the differences are smaller and smaller (definitively smaller than when you introduced it). Now the only difference between the initial migration and a multi step one is that you do not want to update the state in the initial migration.

So why not doing it this way:
- make the initial migration a standard multi step migration.
- make the runner recognize the initial migration and provide it a context that does not do anything to update the state

Reply:
That sounds better, updated